zc_common/remote_resource/filters.py

```diff
@@ -1,55 +1,119 @@
+            pass
+
+        try:
+            field_name = filter_string.split('__')[0]
+            field_filter = filter_class.get_filters().get(field_name, None)
+            is_array_filter = isinstance(field_filter, ArrayFilter)
+            if is_array_filter:
+                filter_value = filter_value.split(',')
+        except AttributeError:
+            pass
+
+        # Allow 'true' or 'false' as values for boolean fields
+        try:
+            if isinstance(queryset.model._meta.get_field(field_name), BooleanField):
+                filter_value = bool(strtobool(filter_value))
+        except FieldDoesNotExist:
+            pass
+
+        filterset_data.append({
+            'field_name': field_name,
+            'field_name_with_lookup': filter_string,
+            'filter_value': filter_value
+        })
+
+        return filterset_data
+
     def filter_queryset(self, request, queryset, view):
         filter_class = self.get_filter_class(view, queryset)
-        primary_key = queryset.model._meta.pk.name
-
-        query_params = {}
 
         for param, value in six.iteritems(request.query_params):
             match = re.search(r'^filter\[(\w+)\]$', param)
             if match:
                 filter_string = match.group(1)
-                field_name = filter_string.split('__').pop(0)
-
-                if field_name not in view.filter_fields.keys():
-                    return queryset.none()
-
-                if len(filter_string) > 1 and field_name == 'id':
-                    filter_string_parts = filter_string.split('__')
-                    filter_string_parts[0] = primary_key
-                    query_params['__'.join(filter_string_parts)] = value
-
-                try:
-                    is_many_to_many_field = isinstance(getattr(queryset.model, filter_string).field, ManyToManyField)
-                    if is_many_to_many_field:
-                        value = value.split(',')
-                except AttributeError:
-                    pass
-
-                # Allow 'true' or 'false' as values for boolean fields
-                try:
-                    if isinstance(queryset.model._meta.get_field(field_name), BooleanField):
-                        value = bool(strtobool(value))
-                except FieldDoesNotExist:
-                    pass
-
-                query_params[filter_string] = value
-
-        if filter_class:
-            return filter_class(query_params, queryset=queryset).qs
+                filters = self._parse_filter_string(queryset, filter_class, filter_string, value)
+
+                for filter_ in filters:
+                    if filter_['field_name'] not in view.filter_fields.keys():
+                        return queryset.none()
+
+                filterset_data = {filter_['field_name_with_lookup']: filter_['filter_value'] for filter_ in filters}
+                if filter_class:
+                    return filter_class(filterset_data, queryset=queryset).qs
```

Comment:
It looks like we have 2 function returns (this one and the `queryset.none()` a few lines above) nested under a for loop over the `request.query_params` - could this result in only applying the first filter to the queryset?

Reply:
I'm actually not 100% sure why we return `queryset.none()` if we can't find the right filter on the view, but since that was explicitly put in there I didn't want to remove it. I think this is the right behavior, if we can't find any of the filters, we should return no results.